File: linked_list/go/linked_list.go
package main

import "fmt"

// Node is the primary block of the linked list
type Node struct {
	value interface{}
	prev  *Node
	next  *Node
}

// LinkedList maintains pointers to the head and tail
type LinkedList struct {
	head *Node
	tail *Node
}

// Insert adds nodes to the linked list
func (list *LinkedList) Insert(value interface{}) *LinkedList {
	node := &Node{value: value}
	if list.head == nil && list.tail == nil {
		list.head = node
		list.tail = node
	} else {
		list.tail.next = node
		node.prev = list.tail
		list.tail = node
	}
	return list
}

func (list LinkedList) String() string {
	str := ""
	str += fmt.Sprint(list.head.value) + " -> "
	str += fmt.Sprint(list.head.next.value) + " -> "
	str += fmt.Sprint(list.head.next.next.value)
	return str
}

func main() {
	list := LinkedList{}
	list.Insert(1).Insert(3).Insert(2)
	fmt.Println(list)
}

File: linked_list/javascript/linked_list.js
function Node(value) {
  this.value = value
  this.next = null
  this.prev = null
}

module.exports = class LinkedList {
  constructor() {
    this.head = null
    this.tail = null
  }

  append(value) {
    const node = new Node(value)
    if (!this.head && !this.tail) {
      this.head = node
      this.tail = node
    }
    else {
      this.tail.next = node
      node.prev = this.tail
      this.tail = node
    }
  }

  prepend(value) {
    const node = new Node(value)
    if (!this.head && !this.tail) {
      this.head = node
      this.tail = node
    }
    else {
      this.head.prev = node
      node.next = this.head
      this.head = node
    }
  }

  insert(value, index = 0) {
    if (index === 0) return this.prepend(value)

    const node = new Node(value)
    var ref = this.head

    while (index-- > 0 && ref) {
      ref = ref.next
    }
    if (!ref) return this.append(value)

    node.prev = ref.prev
    node.next = ref
    ref.prev.next = node
    ref.prev = node
  }

  remove(value) {
    var ref = this.head
    while(ref) {
      if (ref.value === value) {
        ref.prev.next = ref.next
        ref.next.prev = ref.prev
        ref.prev = null
        ref.next = null
        return ref.value
      }
      ref = ref.next
    }
  }
}
